### PR TITLE
Provides caching for the linker

### LINKED ISSUE #478: Links in docstring summary break when the summary is included on another page
```python
"""
Provides L{foo}.
"""

def foo():
    pass
```

image

On that page `foo` links to `#foo` but it should link to `bar.html#foo`.

It's worth noting that ReST links break even worse:

```python
"""
Implements `Foo`_

.. _Foo: http://example.com
"""
```

results in

image

with `` `Foo`_ `` linking to `#id1` for some reason.

### PR DESCRIPTION
**Pull request by tristanlatr**
This PR tries to solve #497, it partially succeeds but we also need to cache the whole Documentable summary stan.  

It does fixes #478

### CODE REVIEW

**Comment by tristanlatr** [comment]
Note: It does not seem to influence the performance of pydoctor, but many warnings are now displayed only once and not many times! so that's good.

**Comment by adiroiban** [comment]
Thanks for the update.

I am a bit busy these days. if @not-my-profile can review this, that would be great :)

File: pydoctor/model.py
Comment (tristanlatr) [inline pydoctor/model.py:388]: The best would be to avoid this workaround cyclic imports. 
In order to do that, we should move all code related to the concrete docstring linkers in a new module: `docstringlinker.py`. I wonder if this change should happen in this PR though ?

**Comment by tristanlatr** [comment]
@adiroiban, I don't think @not-my-profile wants to be involved in pydoctor development. He manifested interest into creating a new doc generator based on mypy internals. So I don't think we should wait for his review.

File: pydoctor/epydoc2stan.py
Comment (adiroiban) [inline pydoctor/epydoc2stan.py:80]: Is this TODO for this PR, or is there a follow-up issue created to track its progress?

File: pydoctor/epydoc2stan.py
Comment (adiroiban) [inline pydoctor/epydoc2stan.py:137]: Maybe have this as docstrig instead of comment?

File: pydoctor/epydoc2stan.py
Comment (adiroiban) [inline pydoctor/epydoc2stan.py:156]: Can we have it as docstring?

File: pydoctor/model.py
Comment (adiroiban) [reply pydoctor/model.py:388]: Better for a separate PR, but also add all this info in a code comment :)

File: pydoctor/test/test_epydoc2stan.py
Comment (adiroiban) [inline pydoctor/test/test_epydoc2stan.py:91]: We might want to enable `black` to create a common style for the code formatting :)
```suggestion
    src = '''
```

**Review by adiroiban — APPROVED** [review]
Hi,Thanks for the update.

If there is no speedup, I am not sure if it's worth it.

Also by not having repeated errors, it might be ok when you run locally, but for CI, I prefer to see all the errors so that I can fix them in one run. 

So.. I am now +1 for this change, but you are the maintainer and if you want to have this code, I am ok with that.

Thanks again!

**Comment by tristanlatr** [comment]
> Also by not having repeated errors, it might be ok when you run locally, but for CI, I prefer to see all the errors so that I can fix them in one run.

We are talking about the same errors again and again for the summary table generation, not different errors. Just run 

```
$ wget -O nodes.py https://sourceforge.net/p/docutils/code/HEAD/tree/tags/docutils-0.17.1/docutils/nodes.py?format=raw
$ time pydoctor nodes.py
```

And you'll see what I mean. See https://github.com/twisted/pydoctor/issues/497

File: pydoctor/epydoc2stan.py
Comment (tristanlatr) [reply pydoctor/epydoc2stan.py:80]: Another PR, It's pretty simple change.

Comment (tristanlatr) [reply pydoctor/epydoc2stan.py:80]: I'll create an issue.

File: pydoctor/epydoc2stan.py
Comment (tristanlatr) [reply pydoctor/epydoc2stan.py:137]: Will do

Comment (tristanlatr) [reply pydoctor/epydoc2stan.py:137]: Actually, I remember. I did not si that not to override the docstirng inherited by `ParsedDocstring`. 
Since this an implementation detail because we're always using the `_CachedEpydocLinker`, and I did not want to change the pretty well documented docstring that we already have for `link_to` and `link_xref`.

**Comment by tristanlatr** [comment]
> If there is no speedup, I am not sure if it's worth it.

@adiroiban,

I've thought to give up this change and instead implement caching for summaries and docstring stan. But here's the deal: there is not way to simply ignore xref warnings when we call `parse_docstring`, so warnings are getting reported when we generate stan for the summary, so at least invalid links present in the fist three lines of docstrings will be reported twice even if we cache the docstring and summary stan (and even if we fix https://github.com/twisted/pydoctor/issues/86. Actually we'de need to have both https://github.com/twisted/pydoctor/issues/86 and https://github.com/twisted/pydoctor/issues/421 fixed and compute the summary document from a version with already resolved links in it). So maybe this patch is just the wrong way of dealing with this problem.

I'm unsure what's the best way forward. The class `_CachedEpydocLinker` is 200 lines of codes and the same in tests. But tests can be recycled anyway. Yes it's a bit of complexity (and multiple invalid link on the same line are reported only once, but it's better than reporting warnings X times the number of summaries we generate for that object. Also if you fix one link in one line in the docstring it's likely that you'll notice the other invalid link that is using the same target on the same line). What I mean is this code can be used until we have fixed https://github.com/twisted/pydoctor/issues/421, https://github.com/twisted/pydoctor/issues/86 and implemented docstring stan caching, then we can remove the `_CachedEpydocLinker` and simply replace all occurences with `_EpydocLinker`.

Please tell me what you think, ok I'm maintainer but I do want to fix issue this the right way.

Thanks.

**Comment by tristanlatr** [comment]
Do you have comments @adiroiban ?
Otherwise we would proceed with the proposed plan.

**Comment by adiroiban** [comment]
I am not -1 on this.
And reducing noise in the output is +1.
So I think that this can be merge.

Thanks!

----------

Sorry... I didn't had time to look into pydoctor for twisted usage and why CI doesn't fail an warnings.

**Comment by tristanlatr** [comment]
Great, thanks!

**Pull request merged**
